File: api/models/inventory.py
"""
Inventory management models - Pool status, lifecycle tracking, and auto-kill
"""

from pydantic import BaseModel
from typing import Optional, Literal
from datetime import datetime
from uuid import UUID


# ===== STATUS TYPES =====

InventoryPoolStatus = Literal["deployed", "warning", "reserve"]
InventoryLifecycleStatus = Literal["active", "incubating", "dead"]
CapacityStatus = Literal["healthy", "warning", "critical"]


# ===== OVERVIEW MODELS =====

class InventoryOverview(BaseModel):
    """Complete inventory overview with pool/lifecycle distribution"""
    workspace_id: UUID

    # Pool distribution
    deployed_count: int = 0
    deployed_target: int = 0
    warning_count: int = 0
    reserve_count: int = 0
    reserve_target: int = 0

    # Lifecycle distribution
    active_count: int = 0
    incubating_count: int = 0
    dead_count: int = 0

    # Death breakdown (separates health issues from business churn)
    killed_by_trigger: int = 0  # System-killed (spam, bounces) - health metric
    deactivated: int = 0  # Manual/business churn - not a health problem

    # Capacity metrics
    total_inboxes: int = 0
    spare_capacity: int = 0
    spare_percentage: float = 0.0
    capacity_status: CapacityStatus = "healthy"

    # Feature flag
    auto_kill_enabled: bool = False

    last_updated: datetime


class InventoryInboxItem(BaseModel):
    """Individual inbox in inventory view"""
    inbox_id: UUID
    email: str
    domain_name: str
    pool_status: Optional[str] = "reserve"
    lifecycle_status: Optional[str] = "incubating"
    age_days: int = 0
    health_score: Optional[int] = None
    hard_bounces_24h: int = 0
    hard_bounces_7d: int = 0
    associated_campaigns: list[str] = []
    warning_reason: Optional[str] = None
    cooldown_ends_at: Optional[datetime] = None


class InventoryInboxListResponse(BaseModel):
    """Response for inbox list endpoint"""
    items: list[InventoryInboxItem]
    total: int
    filtered_by: Optional[str] = None


# ===== AUTO-KILL CONFIGURATION =====

class AutoKillConfig(BaseModel):
    """Auto-kill feature configuration"""
    enabled: bool = False
    cooldown_hours: int = 48
    auto_replace: bool = True
    notify_on_kill: bool = True


class AutoKillConfigResponse(BaseModel):
    """Response after updating auto-kill config"""
    success: bool
    message: str
    config: AutoKillConfig


# ===== KILL AND REPLACE =====

class KillAndReplaceRequest(BaseModel):
    """Request to kill an inbox and optionally replace in campaigns"""
    inbox_id: UUID
    kill_reason: str
    auto_replace: bool = True


class KillAndReplaceResponse(BaseModel):
    """Response after kill and replace operation"""
    killed_inbox_id: UUID
    killed_inbox_email: str
    affected_campaigns: list[str] = []
    replacement_inbox_id: Optional[UUID] = None
    replacement_inbox_email: Optional[str] = None
    success: bool
    message: str


# ===== AUTO-KILL PROCESSING =====

class ProcessAutoKillsResponse(BaseModel):
    """Response after processing auto-kills"""
    processed: int
    message: str
    details: list[dict] = []


# ===== AUDIT LOG =====

class InventoryAuditEvent(BaseModel):
    """Inventory audit log entry"""
    id: UUID
    workspace_id: Optional[UUID]
    inbox_id: Optional[UUID]
    inbox_email: Optional[str]
    event_type: str
    previous_pool_status: Optional[str]
    new_pool_status: Optional[str]
    previous_lifecycle_status: Optional[str]
    new_lifecycle_status: Optional[str]
    reason: Optional[str]
    triggered_by: Optional[str]
    related_campaign_ids: Optional[list[UUID]]
    created_at: datetime


class InventoryAuditLogResponse(BaseModel):
    """Response for audit log endpoint"""
    items: list[InventoryAuditEvent]
    total: int

File: charm-email-os/lib/types/inventory.ts
// ===== INVENTORY STATUS TYPES =====

// Pool status: where inbox is in the inventory flow
export type InventoryPoolStatus = 'deployed' | 'warning' | 'reserve';

// Lifecycle status: maturity of the inbox
export type InventoryLifecycleStatus = 'active' | 'incubating' | 'dead';

// Capacity status: overall inventory health
export type CapacityStatus = 'healthy' | 'warning' | 'critical';


// ===== STATUS COLORS & LABELS =====

export const POOL_STATUS_CONFIG: Record<InventoryPoolStatus, {
  bg: string;
  text: string;
  border: string;
  label: string;
  description: string;
}> = {
  deployed: {
    bg: 'bg-green-500',
    text: 'text-green-700',
    border: 'border-green-500',
    label: 'Deployed',
    description: 'Actively used in campaigns'
  },
  warning: {
    bg: 'bg-yellow-500',
    text: 'text-yellow-700',
    border: 'border-yellow-500',
    label: 'Warning',
    description: 'Needs cooldown due to bounces'
  },
  reserve: {
    bg: 'bg-blue-500',
    text: 'text-blue-700',
    border: 'border-blue-500',
    label: 'Reserve',
    description: 'Ready to join active pool'
  },
};

export const LIFECYCLE_STATUS_CONFIG: Record<InventoryLifecycleStatus, {
  bg: string;
  text: string;
  label: string;
  description: string;
}> = {
  active: {
    bg: 'bg-emerald-100',
    text: 'text-emerald-800',
    label: 'Active',
    description: 'Mature, healthy inbox'
  },
  incubating: {
    bg: 'bg-amber-100',
    text: 'text-amber-800',
    label: 'Incubating',
    description: 'In 2-week warmup period'
  },
  dead: {
    bg: 'bg-red-100',
    text: 'text-red-800',
    label: 'Dead',
    description: 'Killed due to bounces or domain flags'
  },
};

export const CAPACITY_STATUS_CONFIG: Record<CapacityStatus, {
  icon: string;
  color: string;
  bg: string;
  label: string;
}> = {
  healthy: {
    icon: 'CheckCircle',
    color: 'text-green-600',
    bg: 'bg-green-100',
    label: 'Healthy'
  },
  warning: {
    icon: 'AlertTriangle',
    color: 'text-yellow-600',
    bg: 'bg-yellow-100',
    label: 'Warning'
  },
  critical: {
    icon: 'AlertTriangle',
    color: 'text-red-600',
    bg: 'bg-red-100',
    label: 'Critical'
  },
};


// ===== INVENTORY OVERVIEW =====

export interface InventoryOverview {
  workspaceId: string;

  // Pool distribution
  deployedCount: number;
  deployedTarget: number;
  warningCount: number;
  reserveCount: number;
  reserveTarget: number;

  // Lifecycle distribution
  activeCount: number;
  incubatingCount: number;
  deadCount: number;

  // Death breakdown (separates health issues from business churn)
  killedByTrigger: number;  // System-killed (spam, bounces) - health metric
  deactivated: number;  // Manual/business churn - not a health problem

  // Capacity metrics
  totalInboxes: number;
  spareCapacity: number;
  sparePercentage: number;
  capacityStatus: CapacityStatus;

  // Feature flag
  autoKillEnabled: boolean;

  lastUpdated: Date;
}


// ===== INVENTORY INBOX =====

export interface InventoryInbox {
  inboxId: string;
  email: string;
  domainName: string;
  poolStatus: InventoryPoolStatus | null;
  lifecycleStatus: InventoryLifecycleStatus;
  ageDays: number;
  healthScore?: number;
  hardBounces24h: number;
  hardBounces7d: number;
  associatedCampaigns: string[];
  warningReason?: string;
  cooldownEndsAt?: Date;
}

export interface InventoryInboxListResponse {
  items: InventoryInbox[];
  total: number;
  filteredBy?: string;
}


// ===== AUTO-KILL CONFIGURATION =====

export interface AutoKillConfig {
  enabled: boolean;
  cooldownHours: number;
  autoReplace: boolean;
  notifyOnKill: boolean;
}


// ===== KILL AND REPLACE =====

export interface KillAndReplaceRequest {
  inboxId: string;
  killReason: string;
  autoReplace: boolean;
}

export interface KillAndReplaceResponse {
  killedInboxId: string;
  killedInboxEmail: string;
  affectedCampaigns: string[];
  replacementInboxId?: string;
  replacementInboxEmail?: string;
  success: boolean;
  message: string;
}


// ===== AUDIT LOG =====

export interface InventoryAuditEvent {
  id: string;
  workspaceId?: string;
  inboxId?: string;
  inboxEmail?: string;
  eventType: string;
  previousPoolStatus?: string;
  newPoolStatus?: string;
  previousLifecycleStatus?: string;
  newLifecycleStatus?: string;
  reason?: string;
  triggeredBy?: string;
  relatedCampaignIds?: string[];
  createdAt: Date;
}


// ===== BAR CHART DATA =====

export interface InventoryBarChartData {
  // Pool counts
  deployed: number;
  warning: number;
  reserve: number;

  // Total for percentage calculation
  total: number;

  // Targets for reference lines
  deployedTarget: number;
  reserveTarget: number;

  // Capacity
  spareCapacity: number;
  sparePercentage: number;
  capacityStatus: CapacityStatus;
}


// ===== UTILITY FUNCTIONS =====

export function getPoolStatusColor(status: InventoryPoolStatus | null): string {
  if (!status) return 'bg-gray-200';
  return POOL_STATUS_CONFIG[status].bg;
}

export function getLifecycleStatusColor(status: InventoryLifecycleStatus): string {
  return LIFECYCLE_STATUS_CONFIG[status].bg;
}

export function formatCooldownTime(cooldownEndsAt?: Date): string {
  if (!cooldownEndsAt) return '';
  const now = new Date();
  const diff = new Date(cooldownEndsAt).getTime() - now.getTime();
  if (diff <= 0) return 'Cooldown complete';
  const hours = Math.floor(diff / (1000 * 60 * 60));
  const minutes = Math.floor((diff % (1000 * 60 * 60)) / (1000 * 60));
  return `${hours}h ${minutes}m remaining`;
}
